Add NoValInForComprehension syntactic rewrite

gabro opened the pull request:
Closes #153 

I couldn't come up with many creative tests for this, but I'd be glad to add more edge cases if anyone can think of any.

olafurpg commented:
Thank you for this contribution @gabro! You are on fire :) 

I will be away from the laptop for a few days. I will try to review as soon as I get back.

olafurpg commented on scalafix-core/src/main/scala/scalafix/rewrite/NoValInForComprehension.scala:
```diff
@@ -0,0 +1,21 @@
+
+  override def rewrite(ctx: RewriteCtx): Patch = {
+    ctx.tree.collect {
+      case f: Term.ForYield =>
```
I believe this rewrite should also work for `Term.For`. We can make an extractor

```scala
  object TermFor {
    def unapply(tree: Tree): Option[Seq[Enumerator]] = tree match {
      case Term.For(enums, _) => Some(enums)
      case Term.ForYield(enums, _) => Some(enums)
      case _ => None
    }
  }
```

olafurpg replied:
Wait, I don't think we need to match against `Term.For` in the first place. I think it's enough to match directly against `Enumerator.Val`

gabro replied:
Right, I actually though of it, but I couldn't remember whether a `for` without `yield` were valid scala syntax.
Matching `Enumerator.Val` could actually do the trick!

gabro replied:
Oh, ok, I'm ufficially confused by the grammar.

I tried

```scala
(for { x <- List(1, 2, 3) })
```

and the parser complains. By looking at the grammar:


```
`for' (`(' Enumerators `)' | `{' Enumerators `}') {nl} [`yield'] Expr
```

`Expr` is mandatory, so something like


```scala
(for { x <- List(1, 2, 3) } 42)
```

works instead (the type of the expression is `Unit`). I'm wondering why `Expr` is needed without a `yield`...

olafurpg commented on scalafix-tests/src/test/resources/checkSyntax/NoValInForComprehension.source:
```diff
@@ -0,0 +1,21 @@
+    inc2 = inc + 1
+    inc3 = inc2 + 1
+    if inc > 1
+  } yield inc2
```
We should probably add another test for Term.For without `yield`.

gabro commented:
Ok @olafurpg I've simplified the `collect` logic (matching only against `Enumerator.Val` is enough, indeed, nice catch!) and added a test against a `for` without `yield`.

olafurpg commented on scalafix-core/src/main/scala/scalafix/rewrite/NoValInForComprehension.scala:
```diff
@@ -0,0 +1,18 @@
+
+  override def rewrite(ctx: RewriteCtx): Patch = {
+    ctx.tree.collect {
+      case v: Enumerator.Val =>
```
Nice! So small 😄

gabro replied:
small is beautiful 💃
